[shopsys] prepare releaser for work in multiple branches

vitek-rostislav opened the pull request:
| Q             | A
| ------------- | ---
|Description, reason for the PR| Originally, the new version of Shopsys Framework was always released from the current master. But since we started supporting multiple versions (`7.3`, master (`8.0`), ...), we need to tweak the tool to be prepared for these use cases
|New feature| Yes <!-- Do not forget to update docs/ -->
|[BC breaks](https://github.com/shopsys/shopsys/blob/master/docs/contributing/backward-compatibility-promise.md)| No <!-- Do not forget to update UPGRADE.md -->
|Fixes issues| ... <!-- Write "closes #123" for the issue to be closed automatically during merge -->
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

**Important note:** Beside this PR itself, there are other things that need review:

- [x] new Heimdall tool (http://heimdall:8080/view/Tools/job/tool-monorepo-split-branch/) should be reviewed (it was copied from `tool-monorepo-split` and `SPLIT_BRANCH` parameter has been added instead of hardcoded `master` - `CreateAndPushGitTagReleaseWorker` suggests to use this tool when release is performed on other than `master` branch)
- [x] https://github.com/Symplify/ChangelogLinker/pull/3
- [x] https://github.com/Symplify/MonorepoBuilder/pull/4

**TODO:** I asked @PetrHeinz for help as I am missing an answer to the following question:

- [x] How come that Jenkins build is failing after I introduced the dependency on forked `Sympify\ChangelogLinker` in https://github.com/shopsys/shopsys/commit/d8b271c10c75c22bdb342dd9b10b0dcb3a87a07b? (local composer installation works well) - a suggestion of solution [here](https://github.com/shopsys/shopsys/pull/1244#discussion_r309154778) 
- [x] Change the bullet point in the commit message `upgrade files do not use "Unreleased" phrase anymore` - from ~UPGRADE-unreleased.md renamed to UPGRADE-v8.0.0-dev.md~ to `UPGRADE-unreleased.md renamed to UPGRADE-v7.3.2-dev.md`

- [x] Change commit message "CheckChangelogForTodaysDateReleaseWorker now checkouts to "rc-" branch and pushes the changes automatically" (checkout to rc branch has been extracted into a separate commit by fixup)

- [x] Create a proper PR to Symplify current monorepo
- [ ] after this is merged to `7.3`, merge `7.3` branch to `8.0`, then merge `8.0` to `master`, and finally merge `master` to `9.0` branch - beware, `UPGRADE-Unreleased.md` needs to be renamed differently in each of the branches (`UPGRADE-v8.1.0-dev` in master, `UPGRADE-v8.0.1-dev` in `8.0` branch, and `UPGRADE-v9.0.0-dev` in `9.0` branch)

grossmannmartin commented on utils/releaser/src/ReleaseWorker/Release/CheckChangelogForTodaysDateReleaseWorker.php:
```diff
@@ -82,6 +83,7 @@ public function work(Version $version): void
             ));
 
             $this->commit('CHANGELOG.md date updated to today');
+            $this->processRunner->run('git push');
```
Can you please update printed note (`CHANGELOG.md date for "%s" version was updated to "%s".`) and worker description so it's clear this worker pushes?

grossmannmartin replied:
New opinion, please keep pushes manual :) Right now I need to test something and pushed branch by mistake.

vitek-rostislav replied:
updated

vitek-rostislav replied:
fixed

grossmannmartin commented on utils/releaser/src/ReleaseWorker/Release/MergeBranchToMasterReleaseWorker.php:
```diff
@@ -33,7 +33,7 @@ public function getPriority(): int
     public function work(Version $version): void
     {
-        $this->symfonyStyle->note('You need to create a merge commit, see https://github.com/shopsys/shopsys/blob/master/docs/contributing/merging-to-master-on-github.md for detailed instructions.');
+        $this->symfonyStyle->note('You need to create a merge commit, see https://github.com/shopsys/shopsys/blob/master/docs/contributing/merging-on-github.md for detailed instructions.');
```
You forgot to update `CONTRIBUTING.md` file

vitek-rostislav replied:
nice catch, fixed :+1:

grossmannmartin commented on utils/releaser/src/ReleaseWorker/Release/MergeReleaseCandidateBranchReleaseWorker.php:
```diff
@@ -34,7 +34,7 @@ public function getPriority(): int
     {
         $this->symfonyStyle->note('You need to create a merge commit, see https://github.com/shopsys/shopsys/blob/master/docs/contributing/merging-on-github.md for detailed instructions.');
-        $this->confirm('Confirm the release branch was merged to master');
+        $this->confirm(sprintf('Confirm "%s" branch was merged to "%s"', $this->createBranchName($version), $this->initialBranchName));
```
It's problem with this whole stage (release). 
I should run the release stage on `base-branch`? If so, the `CheckChangelogForTodaysDateReleaseWorker` will not behave as expected (unable to find changelog entry as it was not yet merged and push to wrong branch). Running release stage on rc- branch results in this instruction being wrong (`[Manually] Merge "rc-v7-5-0" branch into "rc-v7-5-0"`)

vitek-rostislav replied:
well, the `release` stage is designed this way:
It instructs you to run it on the proper branch, so the `$initialBranchName` is set properly, and then it checkouts to the `rc-xxx-branch` before the changelog is checked, see 3f43513c20ef55fd204f9da8e82ba863410c31d3

grossmannmartin replied:
I saw the checkout, but then it does not come together in my head :)
Can you please extract the checkout into separate release worker? It will be much clearer and we can add another worker if needed.

vitek-rostislav replied:
makes sense. will do it

vitek-rostislav replied:
fixed

vitek-rostislav commented on utils/releaser/src/ReleaseWorker/AbstractShopsysReleaseWorker.php:
```diff
@@ -33,17 +33,23 @@ abstract class AbstractShopsysReleaseWorker implements ReleaseWorkerInterface, S
      * @param \Symfony\Component\Console\Helper\QuestionHelper $questionHelper
      */
-    public function autowire(SymfonyStyle $symfonyStyle, ProcessRunner $processRunner, QuestionHelper $questionHelper): void
+    public function setup(SymfonyStyle $symfonyStyle, ProcessRunner $processRunner, QuestionHelper $questionHelper): void
```
It just suits better from my point of view :wink:

vitek-rostislav commented on utils/releaser/src/ReleaseWorker/AbstractVerifyInitialBranchReleaseWorker.php:
```diff
@@ -0,0 +1,27 @@
+<?php
+
```
fixed

vitek-rostislav commented on utils/releaser/src/ReleaseWorker/AbstractVerifyInitialBranchReleaseWorker.php:
```diff
@@ -0,0 +1,27 @@
+     */
+    public function getDescription(Version $version): string
+    {
+        return 'Verify the release is run on the proper branch';
```
changed

DavidOstrozlik commented on UPGRADE.md:
```diff
@@ -38,7 +38,7 @@ Follow the instructions in the [monorepo upgrade guide](docs/contributing/upgrad
 
 ## Upgrade guides
-* ### [From v7.3.1 to Unreleased](./docs/upgrade/UPGRADE-unreleased.md)
+* ### [From v7.3.1 to v7.3.2-dev](./docs/upgrade/UPGRADE-v7.3.2-dev.md)
```
In commit message you say that you changed this to UPGRADE-v8.0.0-dev.md but its not true. Which version should it be?

vitek-rostislav replied:
yes, it is because I changed the target of the PR during development :slightly_smiling_face: `v7.3.2-dev` is correct, the commit message needs to be changed - I will do it during rebase :wink:

DavidOstrozlik commented on utils/releaser/src/ReleaseWorker/AbstractVerifyInitialBranchReleaseWorker.php:
```diff
@@ -0,0 +1,29 @@
+     */
+    public function getDescription(Version $version): string
+    {
+        return 'Verify you\'re releasing on the proper branch';
```
```suggestion
        return 'Verify that you\'re releasing on the proper branch';
```

DavidOstrozlik commented on composer.json:
```diff
@@ -149,7 +155,7 @@
     "phpstan/phpstan": "^0.11",
     "symplify/easy-coding-standard-tester": "^5.2.17",
-    "symplify/changelog-linker": "^5.2.17",
+    "symplify/changelog-linker": "dev-rv-base-branch",
```
This seems to be problematic on travis. Do you know why?

PetrHeinz replied:
I would publish this fork on Packagist as `shopsys/changelog-linker` in version `v5.5.0`. That should resolve this problem.

DavidOstrozlik commented on utils/releaser/src/FileManipulator/VersionUpgradeFileManipulator.php:
```diff
@@ -12,52 +12,55 @@ final class VersionUpgradeFileManipulator
      * @var string
      */
-    private const FILE_CONTENT_INFORMATION_PATTERN = '#This guide contains instructions to upgrade from version .* to Unreleased#';
+    private const FILE_CONTENT_INFORMATION_PATTERN = '#This guide contains instructions to upgrade from version .* to .*#';
```
Do we really need the `.*` at the end? We know that it should end with `-dev` right?

vitek-rostislav replied:
`-dev` added :slightly_smiling_face:

DavidOstrozlik commented:
Can you please rebase this to 7.3? Releaser now commits lock files and it should be changed to work with other branches as well. Thanks

vitek-rostislav commented:
@DavidOstrozlik rebased :wink:

vitek-rostislav commented:
Hi @DavidOstrozlik, thanks for the testing :+1: 

I have forked the `MonorepoBuilder` package and changed this behavior, see https://github.com/Symplify/MonorepoBuilder/pull/4.

To get this fix in our releaser, I have added a new commit, see 621dea84b
